tinygrad/codegen/lowerer.py

```diff
@@ -34,6 +34,18 @@ def get_grouped_dims(prefix, dims:Tuple[sint, ...], max_sizes:Optional[Tuple[int
           idx //= dims[c]
   return ret[::-1] if reverse else ret
 
+# gets the max of the full shape for each
+# produces a new shape with duplicate reduce dimensions removed
+def get_reduce_dims(reduces:List[List], full_shape:Tuple) -> Tuple[Tuple[int, ...], List[int]]:
+  permute = list(range(len(full_shape)))
+  for ax in range(max(len(r) for r in reduces)):
+    for idx in (idxs:=[r[ax] for r in reduces if ax < len(r)]):
+      permute[idx] = next(i for i in idxs if full_shape[i] == full_shape[idx])
+
+  return functools.reduce(
+    lambda acc, x: ((*acc[0], x[1]), [acc[2] if x[0] == j else j for j in acc[1]], acc[2]+1) if x[0] in permute else acc,
+    enumerate(full_shape), (cast(Tuple[int, ...], tuple()), permute[:], 0))[:2]
```

Comment:
this removes the reused dims from the `full_shape` tuple and the adjusts the permute

it seemed like the best way to do it considering number of lines and code style